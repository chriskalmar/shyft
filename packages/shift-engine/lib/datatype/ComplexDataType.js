

class ComplexDataType {}


export default ComplexDataType


export const isComplexDataType = (obj) => {
  return (obj instanceof ComplexDataType)
}

