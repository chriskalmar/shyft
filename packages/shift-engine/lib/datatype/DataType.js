
import {
  passOrThrow,
  isFunction,
} from '../util';


class DataType {

  constructor (setup = {}) {

    const {
      name,
      description,
      mock,
      validate,
      enforceRequired,
      defaultValue,
      enforceIndex,
    } = setup

    passOrThrow(name, () => 'Missing data type name')
    passOrThrow(description, () => `Missing description for data type '${name}'`)

    passOrThrow(
      isFunction(mock),
      () => `'Missing mock function for data type '${name}'`
    )

    if (validate) {
      passOrThrow(
        isFunction(validate),
        () => `'Invalid validate function for data type '${name}'`
      )

      this.validator = validate
    }

    if (defaultValue) {
      passOrThrow(
        isFunction(defaultValue),
        () => `'Invalid defaultValue function for data type '${name}'`
      )

      this.defaultValue = defaultValue
    }

    this.name = name
    this.description = description
    this.mock = mock

    if (enforceRequired) {
      this.enforceRequired = enforceRequired
    }

    if (enforceIndex) {
      this.enforceIndex = enforceIndex
    }
  }


  validate = (value, context) => {
    if (value && this.validator) {
      this.validator(value, context)
    }
  }


  toString() {
    return this.name
  }

}


export default DataType


export const isDataType = (obj) => {
  return (obj instanceof DataType)
}
