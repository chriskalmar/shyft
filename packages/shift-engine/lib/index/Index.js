
import {
  passOrThrow,
  isArray,
} from '../util';

import _ from 'lodash';

export const INDEX_UNIQUE = 'unique';
export const INDEX_GENERIC = 'generic';
export const indexTypes = [
  INDEX_UNIQUE,
  INDEX_GENERIC,
]


class Index {

  constructor (setup = {}) {

    const {
      type,
      attributes,
    } = setup

    passOrThrow(type, () => 'Missing index type')
    passOrThrow(
      indexTypes.indexOf(type) >= 0,
      () => `Unknown index type '${type}' used, try one of these: '${indexTypes.join(', ')}'`
    )

    passOrThrow(
      isArray(attributes, true),
      () => `Index definition of type '${type}' needs to have a list of attributes`
    )

    attributes.map(attribute => {
      passOrThrow(
        typeof attribute === 'string',
        () => `Index definition of type '${type}' needs to have a list of attribute names`
      )
    })

    passOrThrow(
      attributes.length === _.uniq(attributes).length,
      () => `Index definition of type '${type}' needs to have a list of unique attribute names`
    )


    this.type = type
    this.attributes = attributes

  }


  toString() {
    return this.type
  }

}


export default Index


export const isIndex = (obj) => {
  return (obj instanceof Index)
}


export const processEntityIndexes = (entity, indexes) => {

  passOrThrow(
    isArray(indexes),
    () => `Entity '${entity.name}' indexes definition needs to be an array of indexes`
  )


  const singleAttributeIndexes = []

  const entityAttributes = entity.getAttributes()

  indexes.map((index, idx) => {
    passOrThrow(
      isIndex(index),
      () => `Invalid index definition for entity '${entity.name}' at position '${idx}'`
    )


    index.attributes.map((attributeName) => {

      passOrThrow(
        entityAttributes[ attributeName ],
        () => `Cannot use attribute '${entity.name}.${attributeName}' in index as it does not exist`
      )


      if (index.type === INDEX_UNIQUE) {

        passOrThrow(
          entityAttributes[ attributeName ].required,
          () => `Cannot use attribute '${entity.name}.${attributeName}' in uniqueness index as it is nullable`
        )

        passOrThrow(
          !entityAttributes[ attributeName ].i18n,
          () => `Cannot use attribute '${entity.name}.${attributeName}' in uniqueness index as it is translatable`
        )

        if (index.attributes.length === 1) {
          entityAttributes[ attributeName ].isUnique = true
        }
      }

      if (index.attributes.length === 1) {
        singleAttributeIndexes.push(attributeName)
      }

    })
  })

  // add new index for single attributes with the 'index' flag set but not defined in an index object by the user
  Object.keys(entityAttributes).map(attributeName => {
    if (entityAttributes[attributeName].index) {
      if (!singleAttributeIndexes.includes(attributeName)) {
        indexes.push(new Index({
          type: INDEX_GENERIC,
          attributes: [attributeName]
        }))
      }
    }
  })

  return indexes
}
