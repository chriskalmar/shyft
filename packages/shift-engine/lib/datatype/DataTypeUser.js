
import DataType from './DataType';


export default class DataTypeUser extends DataType {}

export const isDataTypeUser = (obj) => {
  return (obj instanceof DataTypeUser)
}
