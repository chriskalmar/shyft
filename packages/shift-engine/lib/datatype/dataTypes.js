
import DataType from './DataType';
import DataTypeUser from './DataTypeUser';

import casual from 'casual';
import _ from 'lodash';


export const DataTypeUserID = new DataTypeUser({
  name: 'DataTypeUserID',
  description: 'Data type representing a reference to a user',
  mock: /* istanbul ignore next */
    () => casual.integer(2^20, 2^51).toString(),
})



export const DataTypeID = new DataType({
  name: 'DataTypeID',
  description: 'Data type representing unique IDs',
  mock: /* istanbul ignore next */
    () => casual.integer(2^20, 2^51).toString(),
})


export const DataTypeInteger = new DataType({
  name: 'DataTypeInteger',
  description: 'Data type representing integer values',
  mock: /* istanbul ignore next */
    () => casual.integer(-2^10, 2^10),
})


export const DataTypeBigInt = new DataType({
  name: 'DataTypeBigInt',
  description: 'Data type representing big integer values',
  mock: /* istanbul ignore next */
    () => casual.integer(2^20, 2^51).toString(),
})


export const DataTypeFloat = new DataType({
  name: 'DataTypeFloat',
  description: 'Data type representing float values',
  mock: /* istanbul ignore next */
    () => casual.double(-2^10, 2^10),
})


export const DataTypeBoolean = new DataType({
  name: 'DataTypeBoolean',
  description: 'Data type representing boolean values',
  mock: /* istanbul ignore next */
    () => casual.boolean,
})


export const DataTypeString = new DataType({
  name: 'DataTypeString',
  description: 'Data type representing text values',
  mock: /* istanbul ignore next */
    () => casual.title,
})


export const DataTypeJson = new DataType({
  name: 'DataTypeJson',
  description: 'Data type representing json objects',
  mock: /* istanbul ignore next */
    randomJson,
})


export const DataTypeTimestamp = new DataType({
  name: 'DataTypeTimestamp',
  description: 'Data type representing a timestamp',
  mock: /* istanbul ignore next */
    () => new Date(casual.unix_time * 1000),
})


export const DataTypeTimestampTz = new DataType({
  name: 'DataTypeTimestampTz',
  description: 'Data type representing a timestamp with time zone information',
  mock: /* istanbul ignore next */
    () => new Date(casual.unix_time * 1000),
})

export const DataTypeDate = new DataType({
  name: 'DataTypeDate',
  description: 'Data type representing a date',
  mock: /* istanbul ignore next */
    () => new Date(casual.unix_time * 1000),
})


/* istanbul ignore next */
function randomJson() {
  const ret = {}
  const count = casual.integer(1, 5)

  _.times(count, () => {
    const key = _.camelCase( casual.words(2) )
    const value = Math.random() > 0.5
      ? casual.title
      : casual.integer

    ret[ key ] = value
  })

  return ret
}
