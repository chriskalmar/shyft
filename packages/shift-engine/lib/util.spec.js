import Entity from './entity/Entity';
import Index, { INDEX_UNIQUE } from './index/Index';

import {
  passOrThrow,
  resolveFunctionMap,
  isMap,
  isArray,
  mergeMaps,
  mapOverProperties,
  sortDataByKeys,
  processCursor,
  reverseString,
} from './util';

import { DataTypeBoolean, DataTypeString } from './datatype/dataTypes';

describe('util', () => {
  describe('passOrThrow', () => {
    it('should pass on positive condition', () => {
      function fn() {
        passOrThrow(1 < 2, () => 'This error will never happen');
      }

      expect(fn).not.toThrow();
    });

    it('should throw on negative condition', () => {
      function fn() {
        passOrThrow(1 > 2, () => 'This is the very very super important error');
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('should throw if message is not a function', () => {
      function fn() {
        passOrThrow(1 > 0, 'This is the very very super important error');
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });
  });

  describe('resolveFunctionMap', () => {
    const dataMap = {
      a: 123,
      b: 456,
    };

    it('should return the provided map', () => {
      const result = resolveFunctionMap(dataMap);

      expect(result).toEqual(dataMap);
    });

    it('should return the result of the provided function', () => {
      const result = resolveFunctionMap(() => {
        return dataMap;
      });

      expect(result).toEqual(dataMap);
    });
  });

  describe('isMap', () => {
    it('should accept maps', () => {
      expect(isMap({})).toBe(true);
      expect(isMap({ a: 123 })).toBe(true);
      expect(isMap(Object.create({}))).toBe(true);
      expect(isMap(DataTypeBoolean)).toBe(true);
    });

    it('should reject non-maps', () => {
      expect(isMap()).toBe(false);
      expect(isMap(null)).toBe(false);
      expect(isMap(undefined)).toBe(false);
      expect(isMap([])).toBe(false);
      expect(isMap([ 1, 2, 3 ])).toBe(false);
      expect(isMap(() => {})).toBe(false);
      expect(isMap(1234567)).toBe(false);
    });

    it('should reject empty maps if flag `nonEmpty` is set', () => {
      expect(isMap({}, true)).toBe(false);
    });
  });

  describe('isArray', () => {
    it('should accept arrays', () => {
      expect(isArray([])).toBe(true);
      expect(isArray([ 'test' ])).toBe(true);
      expect(isArray([ 1, 2, 3 ])).toBe(true);
    });

    it('should reject non-arrays', () => {
      expect(isArray()).toBe(false);
      expect(isArray(null)).toBe(false);
      expect(isArray(undefined)).toBe(false);
      expect(isArray(Object.create({}))).toBe(false);
      expect(isArray(DataTypeBoolean)).toBe(false);
      expect(isArray(() => {})).toBe(false);
      expect(isArray(1234567)).toBe(false);
    });

    it('should reject empty arrays if flag `nonEmpty` is set', () => {
      expect(isArray([], true)).toBe(false);
    });
  });

  describe('mergeMaps', () => {
    it('should merge 2 maps', () => {
      let obj;

      obj = mergeMaps({}, {});
      expect(obj).toEqual({});

      obj = mergeMaps({ a: 123 }, {});
      expect(obj).toEqual({ a: 123 });

      obj = mergeMaps({}, { b: 456 });
      expect(obj).toEqual({ b: 456 });

      obj = mergeMaps({ a: 123 }, { b: 456 });
      expect(obj).toEqual({ a: 123, b: 456 });

      obj = mergeMaps({ b: 456 }, { a: 123 });
      expect(obj).toEqual({ a: 123, b: 456 });

      obj = mergeMaps({ a: 123, b: 456 }, { b: 789 });
      expect(obj).toEqual({ a: 123, b: 789 });

      obj = mergeMaps(
        { a: 123, b: [ 1, 2, 4 ], c: { deep: [ 7, 8, 9 ] } },
        { b: { b1: 1, b2: 2 } },
      );
      expect(obj).toEqual({
        a: 123,
        b: { b1: 1, b2: 2 },
        c: { deep: [ 7, 8, 9 ] },
      });
    });

    it('should throw if non-maps are provided', () => {
      function fn1() {
        mergeMaps();
      }

      function fn2() {
        mergeMaps({ a: 123 }, []);
      }

      function fn3() {
        mergeMaps('string', {});
      }

      expect(fn1).toThrowErrorMatchingSnapshot();
      expect(fn2).toThrowErrorMatchingSnapshot();
      expect(fn3).toThrowErrorMatchingSnapshot();
    });
  });

  describe('mapOverProperties', () => {
    it('should map over properties and call interatee', () => {
      const keys = [];
      let sum = 0;

      function iteratee(val, key) {
        sum += val;
        keys.push(key);
      }

      const someMap = {
        a: 1,
        b: 4,
        c: 8,
      };

      mapOverProperties(someMap, iteratee);

      expect(keys).toEqual([ 'a', 'b', 'c' ]);
      expect(sum).toBe(13);
    });

    it('should throw if non-maps are provided', () => {
      function fn1() {
        mapOverProperties();
      }

      function fn2() {
        mapOverProperties([]);
      }

      function fn3() {
        mapOverProperties('string');
      }

      expect(fn1).toThrowErrorMatchingSnapshot();
      expect(fn2).toThrowErrorMatchingSnapshot();
      expect(fn3).toThrowErrorMatchingSnapshot();
    });

    it('should throw if iteratee is not a function', () => {
      function fn1() {
        mapOverProperties({});
      }

      function fn2() {
        mapOverProperties({}, []);
      }

      function fn3() {
        mapOverProperties({}, 'string');
      }

      expect(fn1).toThrowErrorMatchingSnapshot();
      expect(fn2).toThrowErrorMatchingSnapshot();
      expect(fn3).toThrowErrorMatchingSnapshot();
    });
  });

  describe('sortDataByKeys', () => {
    it('should return empty result if keys list is empty or invalid', () => {
      const result1 = sortDataByKeys([], { a: 1, b: 3 });
      const result2 = sortDataByKeys({}, { a: 1, b: 3 });
      const result3 = sortDataByKeys(null, { a: 1, b: 3 });

      expect(result1).toEqual([]);
      expect(result2).toEqual([]);
      expect(result3).toEqual([]);
    });

    it('should return null-filled array if data list is empty or invalid', () => {
      const result1 = sortDataByKeys([ 'a', 'b' ], null);
      const result2 = sortDataByKeys([ 'a' ], {});
      const result3 = sortDataByKeys([ 'a', 'b', 'f' ], []);

      expect(result1).toEqual([ null, null ]);
      expect(result2).toEqual([ null ]);
      expect(result3).toEqual([ null, null, null ]);
    });

    it('should sort data based on keys', () => {
      const data = [
        { id: 'a', val: 'lorem' },
        { id: 'b', val: 'ipsum' },
        { id: 'c', val: 'dolor' },
        { id: 'd', val: 'et' },
        { id: 'e', val: 'unde' },
        { id: 'f', val: 'omnis' },
        { id: 'g', val: 'iste' },
        { id: 'a', val: 'natus' },
        { id: 'b', val: 'voluptatem' },
        { id: 'c', val: 'doloremque' },
      ];

      const keys1 = [ 'a', 'c', 'f' ];
      const keys2 = [ 'e', 'e', 'b', 'b', 'c' ];
      const keys3 = [ 'g' ];
      const keys4 = [
        'a-unknown',
        'f',
        'c-unknown',
        'g',
        'c',
        'd-unkown',
        'c',
        'hh',
      ];

      const result1 = sortDataByKeys(keys1, data);
      const result2 = sortDataByKeys(keys2, data);
      const result3 = sortDataByKeys(keys3, data);
      const result4 = sortDataByKeys(keys4, data);

      expect(result1).toEqual([
        { id: 'a', val: 'lorem' },
        { id: 'c', val: 'dolor' },
        { id: 'f', val: 'omnis' },
      ]);

      expect(result2).toEqual([
        { id: 'e', val: 'unde' },
        { id: 'e', val: 'unde' },
        { id: 'b', val: 'ipsum' },
        { id: 'b', val: 'voluptatem' },
        { id: 'c', val: 'dolor' },
      ]);

      expect(result3).toEqual([ { id: 'g', val: 'iste' } ]);

      expect(result4).toEqual([
        null,
        { id: 'f', val: 'omnis' },
        null,
        { id: 'g', val: 'iste' },
        { id: 'c', val: 'dolor' },
        null,
        { id: 'c', val: 'doloremque' },
        null,
      ]);
    });
  });

  describe('processCursor', () => {
    const SomeEntity = new Entity({
      name: 'SomeEntity',
      description: 'Just some description',
      attributes: {
        loginName: {
          type: DataTypeString,
          description: 'Just some description',
          required: true,
        },
        firstName: {
          type: DataTypeString,
          description: 'Just some description',
          required: true,
        },
        lastName: {
          type: DataTypeString,
          description: 'Just some description',
          required: true,
        },
        email: {
          type: DataTypeString,
          description: 'Just some description',
          required: true,
        },
      },
      indexes: [
        new Index({
          type: INDEX_UNIQUE,
          attributes: [ 'loginName' ],
        }),
        new Index({
          type: INDEX_UNIQUE,
          attributes: [ 'firstName', 'lastName' ],
        }),
        new Index({
          type: INDEX_UNIQUE,
          attributes: [ 'email' ],
        }),
      ],
    });

    SomeEntity.getIndexes();

    it('should return empty clause if no cursor provided', () => {
      const result = processCursor();

      expect(result).toEqual({});
    });

    it('should throw if incompatible cursor provided', () => {
      function fn() {
        processCursor({}, { a: 'b' });
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('should throw if cursor is malformed', () => {
      function fn1() {
        processCursor(SomeEntity, { SomeEntity: [ 'b' ] }, []);
      }

      function fn2() {
        processCursor(SomeEntity, { SomeEntity: [ [ {}, {}, {} ] ] }, []);
      }

      function fn3() {
        processCursor(SomeEntity, { SomeEntity: [ {} ] }, []);
      }

      expect(fn1).toThrowErrorMatchingSnapshot();
      expect(fn2).toThrowErrorMatchingSnapshot();
      expect(fn3).toThrowErrorMatchingSnapshot();
    });

    it('should throw if unknown attribute is used', () => {
      function fn() {
        processCursor(
          SomeEntity,
          {
            SomeEntity: [ [ 'iDontKnow', 123 ] ],
          },
          [],
        );
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('should throw if an attribute is used which the data set is not sorted by', () => {
      function fn1() {
        processCursor(SomeEntity, {
          SomeEntity: [ [ 'loginName', 123 ] ],
        });
      }

      function fn2() {
        processCursor(
          SomeEntity,
          {
            SomeEntity: [ [ 'loginName', 123 ] ],
          },
          [
            {
              attribute: 'a',
              direction: 'ASC',
            },
          ],
        );
      }

      function fn3() {
        processCursor(
          SomeEntity,
          {
            SomeEntity: [ [ 'loginName', 123 ], [ 'email', 123 ] ],
          },
          [
            {
              attribute: 'loginName',
              direction: 'ASC',
            },
          ],
        );
      }

      expect(fn1).toThrowErrorMatchingSnapshot();
      expect(fn2).toThrowErrorMatchingSnapshot();
      expect(fn3).toThrowErrorMatchingSnapshot();
    });

    it('should throw if none of the attributes are defined as unique', () => {
      function fn1() {
        processCursor(
          SomeEntity,
          {
            SomeEntity: [ [ 'firstName', 'John' ] ],
          },
          [
            {
              attribute: 'firstName',
              direction: 'ASC',
            },
          ],
        );
      }

      function fn2() {
        processCursor(
          SomeEntity,
          {
            SomeEntity: [ [ 'firstName', 'John' ], [ 'lastName', 'Snow' ] ],
          },
          [
            {
              attribute: 'firstName',
              direction: 'ASC',
            },
            {
              attribute: 'lastName',
              direction: 'ASC',
            },
          ],
        );
      }

      expect(fn1).toThrowErrorMatchingSnapshot();
      expect(fn2).toThrowErrorMatchingSnapshot();
    });

    describe('should return a filter clause based on the provided cursor', () => {
      it('when using attributes that are defined as unique', () => {
        const cursor1 = processCursor(
          SomeEntity,
          {
            SomeEntity: [ [ 'loginName', 'user1' ] ],
          },
          [
            {
              attribute: 'loginName',
              direction: 'ASC',
            },
          ],
        );

        const result1 = {
          loginName: {
            $gt: 'user1',
          },
        };

        const cursor2 = processCursor(
          SomeEntity,
          {
            SomeEntity: [ [ 'loginName', 123 ] ],
          },
          [
            {
              attribute: 'loginName',
              direction: 'DESC',
            },
          ],
        );

        const result2 = {
          loginName: {
            $lt: 123,
          },
        };

        const cursor3 = processCursor(
          SomeEntity,
          {
            SomeEntity: [
              [ 'loginName', 'user1' ],
              [ 'email', 'user1@example.com' ],
            ],
          },
          [
            {
              attribute: 'email',
              direction: 'ASC',
            },
            {
              attribute: 'loginName',
              direction: 'DESC',
            },
          ],
        );

        const result3 = {
          loginName: {
            $lt: 'user1',
          },
        };

        expect(cursor1).toEqual(result1);
        expect(cursor2).toEqual(result2);
        expect(cursor3).toEqual(result3);
      });

      it('when using attributes that are not all defined as unique', () => {
        const row1 = {
          SomeEntity: [ [ 'firstName', 'John' ], [ 'id', 1123 ] ],
        };

        const cursor1 = processCursor(SomeEntity, row1, [
          {
            attribute: 'firstName',
            direction: 'ASC',
          },
          {
            attribute: 'id',
            direction: 'ASC',
          },
        ]);

        const result1 = {
          firstName: {
            $gte: 'John',
          },
          $not: {
            id: {
              $lte: 1123,
            },
            firstName: 'John',
          },
        };

        const cursor2 = processCursor(SomeEntity, row1, [
          {
            attribute: 'firstName',
            direction: 'ASC',
          },
          {
            attribute: 'id',
            direction: 'DESC',
          },
        ]);

        const result2 = {
          firstName: {
            $gte: 'John',
          },
          $not: {
            id: {
              $gte: 1123,
            },
            firstName: 'John',
          },
        };

        const cursor3 = processCursor(SomeEntity, row1, [
          {
            attribute: 'firstName',
            direction: 'DESC',
          },
          {
            attribute: 'id',
            direction: 'DESC',
          },
        ]);

        const result3 = {
          firstName: {
            $lte: 'John',
          },
          $not: {
            id: {
              $gte: 1123,
            },
            firstName: 'John',
          },
        };

        const cursor4 = processCursor(SomeEntity, row1, [
          {
            attribute: 'firstName',
            direction: 'DESC',
          },
          {
            attribute: 'id',
            direction: 'ASC',
          },
        ]);

        const result4 = {
          firstName: {
            $lte: 'John',
          },
          $not: {
            id: {
              $lte: 1123,
            },
            firstName: 'John',
          },
        };

        const row2 = {
          SomeEntity: [
            [ 'firstName', 'John' ],
            [ 'lastName', 'Snow' ],
            [ 'id', 1123 ],
          ],
        };

        const cursor5 = processCursor(SomeEntity, row2, [
          {
            attribute: 'firstName',
            direction: 'ASC',
          },
          {
            attribute: 'lastName',
            direction: 'ASC',
          },
          {
            attribute: 'id',
            direction: 'ASC',
          },
        ]);

        const result5 = {
          firstName: {
            $gte: 'John',
          },
          $not: {
            id: {
              $lte: 1123,
            },
            firstName: 'John',
          },
        };

        const cursor6 = processCursor(SomeEntity, row2, [
          {
            attribute: 'firstName',
            direction: 'ASC',
          },
          {
            attribute: 'lastName',
            direction: 'DESC',
          },
          {
            attribute: 'id',
            direction: 'DESC',
          },
        ]);

        const result6 = {
          firstName: {
            $gte: 'John',
          },
          $not: {
            id: {
              $gte: 1123,
            },
            firstName: 'John',
          },
        };

        const row3 = {
          SomeEntity: [
            [ 'firstName', 'John' ],
            [ 'email', 'john@example.com' ],
            [ 'lastName', 'Snow' ],
            [ 'id', 1123 ],
          ],
        };

        const cursor7 = processCursor(SomeEntity, row3, [
          {
            attribute: 'firstName',
            direction: 'ASC',
          },
          {
            attribute: 'email',
            direction: 'DESC',
          },
          {
            attribute: 'lastName',
            direction: 'DESC',
          },
          {
            attribute: 'id',
            direction: 'DESC',
          },
        ]);

        const result7 = {
          firstName: {
            $gte: 'John',
          },
          $not: {
            email: {
              $gte: 'john@example.com',
            },
            firstName: 'John',
          },
        };

        expect(cursor1).toEqual(result1);
        expect(cursor2).toEqual(result2);
        expect(cursor3).toEqual(result3);
        expect(cursor4).toEqual(result4);
        expect(cursor5).toEqual(result5);
        expect(cursor6).toEqual(result6);
        expect(cursor7).toEqual(result7);
      });

      it('when using in reverse mode', () => {
        const cursor0 = processCursor(
          SomeEntity,
          {
            SomeEntity: [ [ 'loginName', 123 ] ],
          },
          [
            {
              attribute: 'loginName',
              direction: 'DESC',
            },
          ],
          true,
        );

        const result0 = {
          loginName: {
            $gt: 123,
          },
        };

        const row1 = {
          SomeEntity: [ [ 'firstName', 'John' ], [ 'id', 1123 ] ],
        };

        const cursor1 = processCursor(
          SomeEntity,
          row1,
          [
            {
              attribute: 'firstName',
              direction: 'ASC',
            },
            {
              attribute: 'id',
              direction: 'ASC',
            },
          ],
          true,
        );

        const result1 = {
          firstName: {
            $lte: 'John',
          },
          $not: {
            id: {
              $gte: 1123,
            },
            firstName: 'John',
          },
        };

        const cursor2 = processCursor(
          SomeEntity,
          row1,
          [
            {
              attribute: 'firstName',
              direction: 'ASC',
            },
            {
              attribute: 'id',
              direction: 'DESC',
            },
          ],
          true,
        );

        const result2 = {
          firstName: {
            $lte: 'John',
          },
          $not: {
            id: {
              $lte: 1123,
            },
            firstName: 'John',
          },
        };

        const cursor3 = processCursor(
          SomeEntity,
          row1,
          [
            {
              attribute: 'firstName',
              direction: 'DESC',
            },
            {
              attribute: 'id',
              direction: 'DESC',
            },
          ],
          true,
        );

        const result3 = {
          firstName: {
            $gte: 'John',
          },
          $not: {
            id: {
              $lte: 1123,
            },
            firstName: 'John',
          },
        };

        const cursor4 = processCursor(
          SomeEntity,
          row1,
          [
            {
              attribute: 'firstName',
              direction: 'DESC',
            },
            {
              attribute: 'id',
              direction: 'ASC',
            },
          ],
          true,
        );

        const result4 = {
          firstName: {
            $gte: 'John',
          },
          $not: {
            id: {
              $gte: 1123,
            },
            firstName: 'John',
          },
        };

        const row2 = {
          SomeEntity: [
            [ 'firstName', 'John' ],
            [ 'lastName', 'Snow' ],
            [ 'id', 1123 ],
          ],
        };

        const cursor5 = processCursor(
          SomeEntity,
          row2,
          [
            {
              attribute: 'firstName',
              direction: 'ASC',
            },
            {
              attribute: 'lastName',
              direction: 'ASC',
            },
            {
              attribute: 'id',
              direction: 'ASC',
            },
          ],
          true,
        );

        const result5 = {
          firstName: {
            $lte: 'John',
          },
          $not: {
            id: {
              $gte: 1123,
            },
            firstName: 'John',
          },
        };

        const cursor6 = processCursor(
          SomeEntity,
          row2,
          [
            {
              attribute: 'firstName',
              direction: 'ASC',
            },
            {
              attribute: 'lastName',
              direction: 'DESC',
            },
            {
              attribute: 'id',
              direction: 'DESC',
            },
          ],
          true,
        );

        const result6 = {
          firstName: {
            $lte: 'John',
          },
          $not: {
            id: {
              $lte: 1123,
            },
            firstName: 'John',
          },
        };

        const row3 = {
          SomeEntity: [
            [ 'firstName', 'John' ],
            [ 'email', 'john@example.com' ],
            [ 'lastName', 'Snow' ],
            [ 'id', 1123 ],
          ],
        };

        const cursor7 = processCursor(
          SomeEntity,
          row3,
          [
            {
              attribute: 'firstName',
              direction: 'ASC',
            },
            {
              attribute: 'email',
              direction: 'DESC',
            },
            {
              attribute: 'lastName',
              direction: 'DESC',
            },
            {
              attribute: 'id',
              direction: 'DESC',
            },
          ],
          true,
        );

        const result7 = {
          firstName: {
            $lte: 'John',
          },
          $not: {
            email: {
              $lte: 'john@example.com',
            },
            firstName: 'John',
          },
        };

        expect(cursor0).toEqual(result0);
        expect(cursor1).toEqual(result1);
        expect(cursor2).toEqual(result2);
        expect(cursor3).toEqual(result3);
        expect(cursor4).toEqual(result4);
        expect(cursor5).toEqual(result5);
        expect(cursor6).toEqual(result6);
        expect(cursor7).toEqual(result7);
      });
    });
  });

  describe('reverseString', () => {
    it('should reverse a string', () => {
      expect(reverseString('hello')).toEqual('olleh');
      expect(reverseString('')).toEqual('');
      expect(reverseString('a')).toEqual('a');
      expect(reverseString('aBC')).toEqual('CBa');
      expect(reverseString(' x y ')).toEqual(' y x ');
    });
  });
});
