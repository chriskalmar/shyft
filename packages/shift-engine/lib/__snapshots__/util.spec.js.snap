// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util filter processFilterLevel should throw if exact match operators is used with another operator on the same attribute 1`] = `"Cannot combine 'exact match' operator with other operators on attribute 'firstName' used in filter"`;

exports[`util filter processFilterLevel should throw if exact match operators is used with another operator on the same attribute 2`] = `"Cannot combine 'exact match' operator with other operators on attribute 'firstName' used in filter"`;

exports[`util filter processFilterLevel should throw if invalid attributes are used in filter 1`] = `"Unknown attribute name 'something' used in filter"`;

exports[`util filter processFilterLevel should throw if invalid attributes are used in filter 2`] = `"Unknown attribute name 'anything_here' used in filter"`;

exports[`util filter processFilterLevel should throw if invalid attributes are used in filter 3`] = `"Unknown attribute name 'anything_here' used in filter at 'just.here'"`;

exports[`util filter processFilterLevel should throw if invalid operators are used in filter 1`] = `"Unknown or incompatible operator 'ends_with' used on 'isActive' in filter"`;

exports[`util filter processFilterLevel should throw if invalid operators are used in filter 2`] = `"Unknown or incompatible operator 'anything' used on 'firstName' in filter"`;

exports[`util filter processFilterLevel should throw if provided params are invalid 1`] = `"filter needs to be an object of filter criteria"`;

exports[`util filter processFilterLevel should throw if provided params are invalid 2`] = `"optional path in processFilterLevel() needs to be an array"`;

exports[`util filter processFilterLevel should throw if provided params are invalid 3`] = `"filter at 'somewhere' needs to be an object of filter criteria"`;

exports[`util filter processFilterLevel should throw if provided params are invalid 4`] = `"filter at 'somewhere.deeply.nested' needs to be an object of filter criteria"`;

exports[`util filter processFilterLevel should throw if provided params are invalid 5`] = `"processFilterLevel() expects an attribute map"`;

exports[`util filter splitAttributeAndFilterOperator should fail on wrong inputs 1`] = `"invalid filter 'undefined'"`;

exports[`util filter splitAttributeAndFilterOperator should fail on wrong inputs 2`] = `"invalid filter ''"`;

exports[`util filter splitAttributeAndFilterOperator should fail on wrong inputs 3`] = `"invalid filter '[object Object]'"`;

exports[`util filter splitAttributeAndFilterOperator should fail on wrong inputs 4`] = `"invalid filter 'name__'"`;

exports[`util filter splitAttributeAndFilterOperator should fail on wrong inputs 5`] = `"invalid filter 'name___'"`;

exports[`util mapOverProperties should throw if iteratee is not a function 1`] = `"Provided iteratee is not a function"`;

exports[`util mapOverProperties should throw if iteratee is not a function 2`] = `"Provided iteratee is not a function"`;

exports[`util mapOverProperties should throw if iteratee is not a function 3`] = `"Provided iteratee is not a function"`;

exports[`util mapOverProperties should throw if non-maps are provided 1`] = `"Provided object is not a map"`;

exports[`util mapOverProperties should throw if non-maps are provided 2`] = `"Provided object is not a map"`;

exports[`util mapOverProperties should throw if non-maps are provided 3`] = `"Provided object is not a map"`;

exports[`util mergeMaps should throw if non-maps are provided 1`] = `"mergeMaps() expects 2 maps for a merge to work"`;

exports[`util mergeMaps should throw if non-maps are provided 2`] = `"mergeMaps() expects 2 maps for a merge to work"`;

exports[`util mergeMaps should throw if non-maps are provided 3`] = `"mergeMaps() expects 2 maps for a merge to work"`;

exports[`util passOrThrow should throw if message is not a function 1`] = `"passOrThrow() expects messageFn to be a function"`;

exports[`util passOrThrow should throw on negative condition 1`] = `"This is the very very super important error"`;

exports[`util processCursor should throw if an attribute is used which the data set is not sorted by 1`] = `"orderBy needs to be an array of order definitions"`;

exports[`util processCursor should throw if an attribute is used which the data set is not sorted by 2`] = `"Cursor works only on sorted attributes (check: 'loginName')"`;

exports[`util processCursor should throw if an attribute is used which the data set is not sorted by 3`] = `"Cursor works only on sorted attributes (check: 'email')"`;

exports[`util processCursor should throw if cursor is malformed 1`] = `"Cursor malformed"`;

exports[`util processCursor should throw if cursor is malformed 2`] = `"Cursor malformed"`;

exports[`util processCursor should throw if cursor is malformed 3`] = `"Cursor malformed"`;

exports[`util processCursor should throw if incompatible cursor provided 1`] = `"Incompatible cursor for this entity"`;

exports[`util processCursor should throw if none of the attributes are defined as unique 1`] = `"Cursor needs to have at least one attribute defined as unique"`;

exports[`util processCursor should throw if none of the attributes are defined as unique 2`] = `"Cursor needs to have at least one attribute defined as unique"`;

exports[`util processCursor should throw if unknown attribute is used 1`] = `"Unknown attribute 'iDontKnow' used in cursor"`;
